From 250ff042f4e9446b0169a80584d873552bca058c Mon Sep 17 00:00:00 2001
From: Ricardo Cerqueira <cyanogenmod@cerqueira.org>
Date: Wed, 4 Jul 2012 15:26:57 +0100
Subject: [PATCH 1/5] init: Add a mechanism to extend the property service in
 device configs

Instead of every so often adding a new set of stuff here for a specific
device, let the device includes overload the property arrays. The
new device_perms.h header documents it

Change-Id: I477e27abebeb4b85abe5fcaf19333999b1625417
---
 init/device_perms.h     |   29 +++++++++++++++++++++++++++++
 init/property_service.c |    8 ++++++++
 2 files changed, 37 insertions(+)
 create mode 100644 init/device_perms.h

diff --git a/init/device_perms.h b/init/device_perms.h
new file mode 100644
index 0000000..504c0ab
--- /dev/null
+++ b/init/device_perms.h
@@ -0,0 +1,29 @@
+// Overload this file in your own device-specific config if you need
+// non-standard property_perms and/or control_perms structs
+//
+// To avoid conflicts...
+// if you redefine property_perms, #define PROPERTY_PERMS there
+// if you redefine control_perms, #define CONTROL_PARMS there
+//
+// A typical file will look like:
+//
+/*
+
+#define CONTROL_PERMS
+
+struct {
+    const char *service;
+    unsigned int uid;
+    unsigned int gid;
+} control_perms[] = {
+    // The default perms
+    { "dumpstate",AID_SHELL, AID_LOG },
+    { "ril-daemon",AID_RADIO, AID_RADIO },
+    // My magic device perms
+    { "rawip_vsnet1",AID_RADIO, AID_RADIO },
+    { "rawip_vsnet2",AID_RADIO, AID_RADIO },
+    { "rawip_vsnet3",AID_RADIO, AID_RADIO },
+    { "rawip_vsnet4",AID_RADIO, AID_RADIO },
+     {NULL, 0, 0 }
+};
+*/
diff --git a/init/property_service.c b/init/property_service.c
index b844f67..92ff851 100644
--- a/init/property_service.c
+++ b/init/property_service.c
@@ -68,6 +68,8 @@
 #include "util.h"
 #include "log.h"
 
+#include <device_perms.h>
+
 #define PERSISTENT_PROPERTY_DIR  "/data/property"
 
 static int persistent_properties_loaded = 0;
@@ -76,6 +78,7 @@ static int property_area_inited = 0;
 static int property_set_fd = -1;
 
 /* White list of permissions for setting property services. */
+#ifndef PROPERTY_PERMS
 struct {
     const char *prefix;
     unsigned int uid;
@@ -124,11 +127,14 @@ struct {
     { "usb_uicc.", AID_SYSTEM,  0 },
     { NULL, 0, 0 }
 };
+/* Avoid extending this array. Check device_perms.h */
+#endif
 
 /*
  * White list of UID that are allowed to start/stop services.
  * Currently there are no user apps that require.
  */
+#ifndef CONTROL_PERMS
 struct {
     const char *service;
     unsigned int uid;
@@ -138,6 +144,8 @@ struct {
     { "ril-daemon",AID_RADIO, AID_RADIO },
      {NULL, 0, 0 }
 };
+/* Avoid extending this array. Check device_perms.h */
+#endif
 
 typedef struct {
     size_t size;
-- 
1.7.9.5

